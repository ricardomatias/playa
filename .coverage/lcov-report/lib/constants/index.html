<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for lib/constants</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> lib/constants
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">87.5% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>56/64</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">66.67% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>4/6</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">33.33% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>2/6</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">88.89% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>56/63</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line high'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="chords.js"><a href="chords.js.html">chords.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="36" class="abs high">36/36</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="36" class="abs high">36/36</td>
	</tr>

<tr>
	<td class="file empty" data-value="index.js"><a href="index.js.html">index.js</a></td>
	<td data-value="0" class="pic empty"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	</tr>

<tr>
	<td class="file high" data-value="intervals.js"><a href="intervals.js.html">intervals.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	</tr>

<tr>
	<td class="file high" data-value="note.js"><a href="note.js.html">note.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	</tr>

<tr>
	<td class="file high" data-value="rhythmsDistributions.js"><a href="rhythmsDistributions.js.html">rhythmsDistributions.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	</tr>

<tr>
	<td class="file high" data-value="scales.js"><a href="scales.js.html">scales.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	</tr>

<tr>
	<td class="file medium" data-value="ticks.js"><a href="ticks.js.html">ticks.js</a></td>
	<td data-value="60" class="pic medium"><div class="chart"><div class="cover-fill" style="width: 60%;"></div><div class="cover-empty" style="width:40%;"></div></div></td>
	<td data-value="60" class="pct medium">60%</td>
	<td data-value="20" class="abs medium">12/20</td>
	<td data-value="66.67" class="pct medium">66.67%</td>
	<td data-value="6" class="abs medium">4/6</td>
	<td data-value="33.33" class="pct low">33.33%</td>
	<td data-value="6" class="abs low">2/6</td>
	<td data-value="63.16" class="pct medium">63.16%</td>
	<td data-value="19" class="abs medium">12/19</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Wed Feb 20 2019 14:43:58 GMT+0100 (Central European Standard Time)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
<script src="../../block-navigation.js"></script>
</body>
</html>
