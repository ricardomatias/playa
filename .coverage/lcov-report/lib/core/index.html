<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for lib/core</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> lib/core
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">88.17% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>343/389</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">81.98% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>141/172</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">83.82% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>57/68</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">87.83% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>332/378</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line high'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="Chord.js"><a href="Chord.js.html">Chord.js</a></td>
	<td data-value="96.9" class="pic high"><div class="chart"><div class="cover-fill" style="width: 96%;"></div><div class="cover-empty" style="width:4%;"></div></div></td>
	<td data-value="96.9" class="pct high">96.9%</td>
	<td data-value="129" class="abs high">125/129</td>
	<td data-value="90.91" class="pct high">90.91%</td>
	<td data-value="44" class="abs high">40/44</td>
	<td data-value="94.44" class="pct high">94.44%</td>
	<td data-value="18" class="abs high">17/18</td>
	<td data-value="96.77" class="pct high">96.77%</td>
	<td data-value="124" class="abs high">120/124</td>
	</tr>

<tr>
	<td class="file high" data-value="HarmonyBase.js"><a href="HarmonyBase.js.html">HarmonyBase.js</a></td>
	<td data-value="84.62" class="pic high"><div class="chart"><div class="cover-fill" style="width: 84%;"></div><div class="cover-empty" style="width:16%;"></div></div></td>
	<td data-value="84.62" class="pct high">84.62%</td>
	<td data-value="13" class="abs high">11/13</td>
	<td data-value="66.67" class="pct medium">66.67%</td>
	<td data-value="6" class="abs medium">4/6</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="84.62" class="pct high">84.62%</td>
	<td data-value="13" class="abs high">11/13</td>
	</tr>

<tr>
	<td class="file high" data-value="Key.js"><a href="Key.js.html">Key.js</a></td>
	<td data-value="94.67" class="pic high"><div class="chart"><div class="cover-fill" style="width: 94%;"></div><div class="cover-empty" style="width:6%;"></div></div></td>
	<td data-value="94.67" class="pct high">94.67%</td>
	<td data-value="75" class="abs high">71/75</td>
	<td data-value="83.33" class="pct high">83.33%</td>
	<td data-value="36" class="abs high">30/36</td>
	<td data-value="92.31" class="pct high">92.31%</td>
	<td data-value="13" class="abs high">12/13</td>
	<td data-value="94.67" class="pct high">94.67%</td>
	<td data-value="75" class="abs high">71/75</td>
	</tr>

<tr>
	<td class="file low" data-value="Mode.js"><a href="Mode.js.html">Mode.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="24" class="abs low">0/24</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="9" class="abs low">0/9</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="24" class="abs low">0/24</td>
	</tr>

<tr>
	<td class="file high" data-value="Note.js"><a href="Note.js.html">Note.js</a></td>
	<td data-value="89.74" class="pic high"><div class="chart"><div class="cover-fill" style="width: 89%;"></div><div class="cover-empty" style="width:11%;"></div></div></td>
	<td data-value="89.74" class="pct high">89.74%</td>
	<td data-value="78" class="abs high">70/78</td>
	<td data-value="90.48" class="pct high">90.48%</td>
	<td data-value="42" class="abs high">38/42</td>
	<td data-value="84.21" class="pct high">84.21%</td>
	<td data-value="19" class="abs high">16/19</td>
	<td data-value="89.61" class="pct high">89.61%</td>
	<td data-value="77" class="abs high">69/77</td>
	</tr>

<tr>
	<td class="file high" data-value="Scale.js"><a href="Scale.js.html">Scale.js</a></td>
	<td data-value="93.94" class="pic high"><div class="chart"><div class="cover-fill" style="width: 93%;"></div><div class="cover-empty" style="width:7%;"></div></div></td>
	<td data-value="93.94" class="pct high">93.94%</td>
	<td data-value="66" class="abs high">62/66</td>
	<td data-value="82.86" class="pct high">82.86%</td>
	<td data-value="35" class="abs high">29/35</td>
	<td data-value="80" class="pct high">80%</td>
	<td data-value="10" class="abs high">8/10</td>
	<td data-value="93.44" class="pct high">93.44%</td>
	<td data-value="61" class="abs high">57/61</td>
	</tr>

<tr>
	<td class="file empty" data-value="index.js"><a href="index.js.html">index.js</a></td>
	<td data-value="0" class="pic empty"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	</tr>

<tr>
	<td class="file high" data-value="types.js"><a href="types.js.html">types.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Wed Feb 20 2019 15:09:45 GMT+0100 (Central European Standard Time)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
<script src="../../block-navigation.js"></script>
</body>
</html>
