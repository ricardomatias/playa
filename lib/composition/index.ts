export { createMotif } from './motif';
export { createChordProgression } from './chord-progression';
export { createMelodies } from './melodies';
export { createPercussion } from './percussion';
export { createArp, genHarmonicShift } from './arp';

export * as Rhythm from './rhythm';
export * from './sequence';
export * from './movement';
