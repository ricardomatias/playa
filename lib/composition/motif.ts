import { choose } from '../tools';
import { Note } from '../core/Note';
import { Time, TimeFormat } from '../core/Time';
import { NoteEvent } from '../core/NoteEvent';
import { expandDuration, isEvent, mapStartToEvent } from '../tools/event';
import { Event } from '../core/Event';

/**
 * Generates a motif
 *
 * @function createMotif
 * @memberof Composition
 *
 * @example
 * createMotif(new Scale('A', Scale.MAJOR).notes, Rhythm.free('1:0:0', ['4n', '8n', '16n']))
 *
 * @param {Array<Note>} notes
 * @param {Array<String>} rhythm note durations
 * @param {TimeFormat} [startTime = 0] when to start the motif
 * @return {Array<NoteEvent>}
 */
export function createMotif(notes: Note[], rhythm: string[] | Event[], startTime: TimeFormat = 0): NoteEvent[] {
	if (!isEvent(rhythm)) {
		rhythm = expandDuration(rhythm);
	}

	const start = new Time(startTime);
	const pattern = rhythm.map((event: Event) => {
		const note = choose(notes);

		if (event.isRest) {
			return NoteEvent(mapStartToEvent(start, event));
		}

		return NoteEvent({
			...mapStartToEvent(start, event),
			note: note.n,
			midi: note.m,
		});
	});

	return pattern;
}

