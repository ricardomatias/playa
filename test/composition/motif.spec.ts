import * as R from 'ramda';
import { Scale, Chord } from '../../lib/core';
import { createMotif } from '../../lib/composition/motif';
import { distribute } from '@ricardomatias/roll';
import { chooseMany } from '../../lib/tools/choose';
import * as Rhythm from '../../lib/composition/rhythm';
import random from '../../lib/tools/random';
import { Ticks } from '../../lib/constants';
import '../matchers';

const { Mixed } = Rhythm.Presets;

const ONE_BAR = Ticks['1n'];

describe('A Motif test suite', () => {
	it('should generate motif based on scale', () => {
		// given
		random.setSeed('test');

		const scale = new Scale('A', Scale.Major);

		let rhythm = Rhythm.free('1:0:0', Mixed, [], distribute.decreasing);

		// when
		let motif = createMotif(scale.notes, rhythm);

		// then
		expect(motif).toMatchSnapshot();
		expect(motif).toLastAround(ONE_BAR);

		// when
		random.setSeed('test2');

		rhythm = Rhythm.free('2:0:0', Mixed, [], distribute.decreasing);

		motif = createMotif(scale.notes, rhythm);

		// then
		expect(motif).toMatchSnapshot();
		expect(motif).toLastAround(ONE_BAR * 2);
	});

	it('should generate createMotif based on chord', () => {
		// given
		random.setSeed('test');

		const chord = Chord.fromName('Dbm6');
		let rhythm = Rhythm.free('1:0:0', Mixed, [], distribute.decreasing);

		// when
		let motif = createMotif(chord.notes, rhythm);

		expect(motif).toMatchSnapshot();
		expect(motif).toLastAround(ONE_BAR);

		// when
		random.setSeed('test2');

		rhythm = Rhythm.free('2:0:0', Mixed, [], distribute.decreasing);

		motif = createMotif(chord.notes, rhythm);

		expect(motif).toMatchSnapshot();
		expect(motif).toLastAround(ONE_BAR * 2);
	});

	it('should return pattern with time = start time', () => {
		// given
		random.setSeed('test');

		const chord = Chord.fromName('Dbm6');
		const rhythm = Rhythm.free('0:3:0', Mixed);

		const motif = createMotif(chord.notes, rhythm, 480);

		expect(motif).toMatchSnapshot();
	});

	it('should generate motif with rests', () => {
		// given
		random.setSeed('test');

		const scale = new Scale('A', Scale.Major);
		const durations = chooseMany(Mixed, random.int(R.length(Mixed)));

		const rhythm = Rhythm.free('1:0:0', Mixed, durations, distribute.decreasing);

		// when
		const motif = createMotif(scale.notes, rhythm);

		// then
		expect(motif).toMatchSnapshot();
		expect(motif).toLastAround(ONE_BAR);
	});

	it('should generate motif with given durations', () => {
		// given
		random.setSeed('test');

		const scale = new Scale('A', Scale.Major);

		// when
		const motif = createMotif(scale.notes, [ '4n', '2n', '8nd' ]);

		// then
		expect(motif).toMatchSnapshot();
		expect(motif).toLastAround(1800);
	});
});
