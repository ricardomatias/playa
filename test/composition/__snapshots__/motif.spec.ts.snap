// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A Motif test suite should generate createMotif based on chord 1`] = `
[
  {
    "dur": 240,
    "isRest": false,
    "midi": 64,
    "next": 240,
    "note": "E3",
    "time": 0,
  },
  {
    "dur": 480,
    "isRest": false,
    "midi": 68,
    "next": 720,
    "note": "Ab3",
    "time": 240,
  },
  {
    "dur": 480,
    "isRest": false,
    "midi": 70,
    "next": 1200,
    "note": "Bb3",
    "time": 720,
  },
  {
    "dur": 240,
    "isRest": false,
    "midi": 64,
    "next": 1440,
    "note": "E3",
    "time": 1200,
  },
  {
    "dur": 320,
    "isRest": false,
    "midi": 68,
    "next": 1760,
    "note": "Ab3",
    "time": 1440,
  },
  {
    "dur": 160,
    "isRest": false,
    "midi": 64,
    "next": 1920,
    "note": "E3",
    "time": 1760,
  },
]
`;

exports[`A Motif test suite should generate createMotif based on chord 2`] = `
[
  {
    "dur": 240,
    "isRest": false,
    "midi": 61,
    "next": 240,
    "note": "Db3",
    "time": 0,
  },
  {
    "dur": 720,
    "isRest": false,
    "midi": 68,
    "next": 960,
    "note": "Ab3",
    "time": 240,
  },
  {
    "dur": 480,
    "isRest": false,
    "midi": 68,
    "next": 1440,
    "note": "Ab3",
    "time": 960,
  },
  {
    "dur": 240,
    "isRest": false,
    "midi": 70,
    "next": 1680,
    "note": "Bb3",
    "time": 1440,
  },
  {
    "dur": 720,
    "isRest": false,
    "midi": 68,
    "next": 2400,
    "note": "Ab3",
    "time": 1680,
  },
  {
    "dur": 720,
    "isRest": false,
    "midi": 64,
    "next": 3120,
    "note": "E3",
    "time": 2400,
  },
  {
    "dur": 240,
    "isRest": false,
    "midi": 64,
    "next": 3360,
    "note": "E3",
    "time": 3120,
  },
  {
    "dur": 240,
    "isRest": false,
    "midi": 68,
    "next": 3600,
    "note": "Ab3",
    "time": 3360,
  },
  {
    "dur": 240,
    "isRest": false,
    "midi": 64,
    "next": 3840,
    "note": "E3",
    "time": 3600,
  },
]
`;

exports[`A Motif test suite should generate motif based on scale 1`] = `
[
  {
    "dur": 240,
    "isRest": false,
    "midi": 74,
    "next": 240,
    "note": "D4",
    "time": 0,
  },
  {
    "dur": 480,
    "isRest": false,
    "midi": 74,
    "next": 720,
    "note": "D4",
    "time": 240,
  },
  {
    "dur": 480,
    "isRest": false,
    "midi": 80,
    "next": 1200,
    "note": "G#4",
    "time": 720,
  },
  {
    "dur": 240,
    "isRest": false,
    "midi": 71,
    "next": 1440,
    "note": "B3",
    "time": 1200,
  },
  {
    "dur": 320,
    "isRest": false,
    "midi": 74,
    "next": 1760,
    "note": "D4",
    "time": 1440,
  },
  {
    "dur": 160,
    "isRest": false,
    "midi": 73,
    "next": 1920,
    "note": "C#4",
    "time": 1760,
  },
]
`;

exports[`A Motif test suite should generate motif based on scale 2`] = `
[
  {
    "dur": 240,
    "isRest": false,
    "midi": 69,
    "next": 240,
    "note": "A3",
    "time": 0,
  },
  {
    "dur": 720,
    "isRest": false,
    "midi": 74,
    "next": 960,
    "note": "D4",
    "time": 240,
  },
  {
    "dur": 480,
    "isRest": false,
    "midi": 74,
    "next": 1440,
    "note": "D4",
    "time": 960,
  },
  {
    "dur": 240,
    "isRest": false,
    "midi": 78,
    "next": 1680,
    "note": "F#4",
    "time": 1440,
  },
  {
    "dur": 720,
    "isRest": false,
    "midi": 78,
    "next": 2400,
    "note": "F#4",
    "time": 1680,
  },
  {
    "dur": 720,
    "isRest": false,
    "midi": 73,
    "next": 3120,
    "note": "C#4",
    "time": 2400,
  },
  {
    "dur": 240,
    "isRest": false,
    "midi": 73,
    "next": 3360,
    "note": "C#4",
    "time": 3120,
  },
  {
    "dur": 240,
    "isRest": false,
    "midi": 78,
    "next": 3600,
    "note": "F#4",
    "time": 3360,
  },
  {
    "dur": 240,
    "isRest": false,
    "midi": 73,
    "next": 3840,
    "note": "C#4",
    "time": 3600,
  },
]
`;

exports[`A Motif test suite should generate motif with given durations 1`] = `
[
  {
    "dur": 480,
    "isRest": false,
    "midi": 74,
    "next": 480,
    "note": "D4",
    "time": 0,
  },
  {
    "dur": 960,
    "isRest": false,
    "midi": 69,
    "next": 1440,
    "note": "A3",
    "time": 480,
  },
  {
    "dur": 360,
    "isRest": false,
    "midi": 71,
    "next": 1800,
    "note": "B3",
    "time": 1440,
  },
]
`;

exports[`A Motif test suite should generate motif with rests 1`] = `
[
  {
    "dur": 160,
    "isRest": true,
    "midi": -1,
    "next": 160,
    "note": "",
    "time": 0,
  },
  {
    "dur": 320,
    "isRest": false,
    "midi": 76,
    "next": 480,
    "note": "E4",
    "time": 160,
  },
  {
    "dur": 160,
    "isRest": false,
    "midi": 73,
    "next": 640,
    "note": "C#4",
    "time": 480,
  },
  {
    "dur": 320,
    "isRest": true,
    "midi": -1,
    "next": 960,
    "note": "",
    "time": 640,
  },
  {
    "dur": 480,
    "isRest": false,
    "midi": 78,
    "next": 1440,
    "note": "F#4",
    "time": 960,
  },
  {
    "dur": 160,
    "isRest": false,
    "midi": 73,
    "next": 1600,
    "note": "C#4",
    "time": 1440,
  },
  {
    "dur": 80,
    "isRest": true,
    "midi": -1,
    "next": 1680,
    "note": "",
    "time": 1600,
  },
  {
    "dur": 480,
    "isRest": false,
    "midi": 73,
    "next": 2160,
    "note": "C#4",
    "time": 1680,
  },
]
`;

exports[`A Motif test suite should return pattern with time = start time 1`] = `
[
  {
    "dur": 720,
    "isRest": false,
    "midi": 68,
    "next": 1200,
    "note": "Ab3",
    "time": 480,
  },
  {
    "dur": 480,
    "isRest": false,
    "midi": 68,
    "next": 1680,
    "note": "Ab3",
    "time": 1200,
  },
  {
    "dur": 240,
    "isRest": false,
    "midi": 68,
    "next": 1920,
    "note": "Ab3",
    "time": 1680,
  },
]
`;
